playground.js

```diff
@@ -0,0 +1,278 @@
+      content: 'The easiest way to call the Flex Marketplace API using the JS SDK to try things out, to learn the APIs and to test your ideas.'
+    },
+    {
+      header: 'Options',
+      optionList: optionDefinitions
+    }
+  ]));
+};
+
+const printResponse = response => {
+  console.log(JSON.stringify(response.data, null, 4));
+  return response;
+}
+
+const apiDocs = () =>
+      open('https://www.sharetribe.com/api-reference/marketplace.html?javascript#marketplace-api')
+      .then(() => '');
+
+// Parse command line args
+//
+const options = commandLineArgs(optionDefinitions);
+
+if (options.apidocs) {
+  apiDocs();
+  process.exit();
+}
+
+if (options.help || (!options.raw && options.clientid == null)) {
+  printUsage();
+  process.exit();
+};
+
+let scriptSrc;
+if (options.script) {
+  scriptSrc = fs.createReadStream(options.script);
+} else {
+  scriptSrc = null;
+}
+
+const setupContext = (ctx, sdk) => {
+  ctx.sharetribeSdk = sharetribeSdk;
+  if (sdk) {
+    ctx.sdk = sdk;
+  }
+  ctx.printResponse = printResponse;
+  ctx.apiDocs = apiDocs;
+  ctx.LatLng = LatLng;
+  ctx.LatLngBounds = LatLgnBounds;
+  ctx.UUID = UUID;
+  ctx.Money = Money;
+  ctx.BigDecimal = BigDecimal;
+}
+
+const startRepl = (sdk, rawMode, src) =>
+      () => {
+        if (src) {
+          // Setup context
+          setupContext(global, sdk);
+
+          // Start REPL
+          console.log('Executing script...');
+          console.log();
+          repl.start({
+            prompt: '> ',
+            input: src,
+            output: process.stdout,
+            useGlobal: true,
+            writer: result => result instanceof Promise ? '' : util.inspect(result)
+          });
+        } else {
+          printWelcomeMessage(rawMode);
+          // Start REPL
+          const replInstance = repl.start('> ');
+
+          // Attach history
+          replInstance.setupHistory('./.repl_history', () => null);
+
+          // Setup context
+          const ctx = replInstance.context;
+          setupContext(ctx, sdk);
+        }
+      }
+
+const exitOnFailure = msg =>
+      () => {
+        console.error(msg)
+        process.exit(1);
+      };
+
+// If client ID is provided on start, instantiate and expose sdk
+if (options.clientid) {
+  const sdk = sharetribeSdk.createInstance({
+    clientId: options.clientid,
+    tokenStore: sharetribeSdk.tokenStore.memoryStore()
+  });
+
+  // If user auth info is given too, log in with the user.
+  if (options.user && options.password) {
+    console.log(`Initializing SDK instance with Client ID: ${options.clientid}...`)
+    sdk.marketplace.show()
+      .then(result => console.log(`Successfully connected to ${result.data.data.attributes.name} marketplace.`))
+      .catch(exitOnFailure(`Unable to access the Marketplace API with the given Client ID: ${options.clientid}.`))
+      .then(() => {
+        console.log(`Logging in user ${options.user}...`);
+        sdk.login({
```

Comment:
Comment: What about returning the sdk.login promise, and moving the "catch" and "then" outside of this handler? In addition, does the error message contain enough information to know what went wrong? If yes, what about structuring the promise chain something like this:

```js
sdk.marketplace.show()
   .then(() => {
     console.log(`Successfully connected to ${result.data.data.attributes.name} marketplace.`)
     console.log(`Logging in user ${options.user}...`);

     return sdk.login({ ... })
   })
   .then(startRepl(sdk, false, scriptSrc))
   .catch((e) => {
     if (marketplaceShowFailed) {
       console.error(`Unable to access the Marketplace API with the given Client ID: ${options.clientid}.`) 
     }

     if (loginFailed) {
       console.error(`Unable to log in with the email: ${options.user} and password: ${options.password}`)
     }
   })
```

Reply:
What you said about returning promise from login was my first attempt but I didn't realize I need an explicit return, couldn't get it work and then fell back to this version. So I'll give it a go to switch to cleaner chain now again. Regarding just one catch block; I don't think I easily have enough information to do that so it's just way easier to have a small number of catch blocks interposed to do the right thing. I'll stick with that pattern.

Reply:
Yeah, here you have the luxury of `process.exit()` which you don't normally have :) That's why the pattern works in this case. But in normal cases you'd have to do what I suggested, because you can't escape the Promise chain with `process.exit()` and even throwing doesn't help, because the Promise chain catches exceptions.